File: riistaandroid/src/main/java/fi/riista/mobile/pages/MyDetailsFragment.java
package fi.riista.mobile.pages;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Map;

import fi.riista.mobile.R;
import fi.riista.mobile.activity.BaseActivity;
import fi.riista.mobile.models.user.Address;
import fi.riista.mobile.models.user.Occupation;
import fi.riista.mobile.models.user.Rhy;
import fi.riista.mobile.models.user.UserInfo;
import fi.riista.mobile.utils.AppPreferences;

/**
 * My details/hunting card page.
 * Simply displays stored user information.
 */
public class MyDetailsFragment extends PageFragment {

    private static final String SIMPLE_DATE_FORMAT = "dd.MM.yyyy";
    private static final String SIMPLE_DURATION_FORMAT = "%s - %s";
    private static final String ADDRESS_FORMAT = "%s\n%s %s\n%s";
    private static final String MEMBERSHIP_NAME_FORMAT = "%s (%s)";
    private static final String OCCUPATION_FORMAT = "%s\n%s\n%s";

    private Context mContext;
    private TextView mNameText;
    private TextView mDateOfBirthText;
    private TextView mHomeMunicipality;
    private TextView mAddressText;
    private TextView mHunterIdText;
    private TextView mPaymentText;
    private TextView mMembershipText;
    private TextView mInsuranceText;
    private TextView mNoValidLicenceItem;

    private View mHuntingBanItem;
    private TextView mHuntingBanText;

    private View mHunterIdItem;
    private View mPaymentItem;
    private View mMembershipItem;

    private View mOccupationsTitle;
    private ViewGroup mOccupationsContainer;

    public static MyDetailsFragment newInstance() {
        return new MyDetailsFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_my_details, container, false);
        mNameText = (TextView) view.findViewById(R.id.my_details_name_value);
        mDateOfBirthText = (TextView) view.findViewById(R.id.my_details_date_of_birth_value);
        mHomeMunicipality = (TextView) view.findViewById(R.id.my_details_home_municipality_value);
        mAddressText = (TextView) view.findViewById(R.id.my_details_address_value);
        mHunterIdText = (TextView) view.findViewById(R.id.my_details_hunter_id_value);
        mPaymentText = (TextView) view.findViewById(R.id.my_details_payment_value);
        mMembershipText = (TextView) view.findViewById(R.id.my_details_membership_value);
        mInsuranceText = (TextView) view.findViewById(R.id.my_details_insurance_policy);
        mNoValidLicenceItem = (TextView) view.findViewById(R.id.my_details_no_valid_license);

        mHuntingBanItem = view.findViewById(R.id.my_details_hunting_ban_item);
        mHuntingBanText = (TextView) view.findViewById(R.id.my_details_hunting_ban_value);

        mHunterIdItem = view.findViewById(R.id.my_details_hunter_id_item);
        mPaymentItem = view.findViewById(R.id.my_details_payment_item);
        mMembershipItem = view.findViewById(R.id.my_details_membership_item);

        mOccupationsTitle = view.findViewById(R.id.my_details_occupations_title);
        mOccupationsContainer = (ViewGroup) view.findViewById(R.id.my_details_occupations_container);

        mContext = inflater.getContext();

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();

        setViewTitle(getString(R.string.title_my_details));
        ((BaseActivity) getActivity()).onHasActionbarMenu(false);

        UserInfo info = AppPreferences.getUserInfo(mContext);

        // Timestamp check is to make sure the data from server is not in outdated format.
        if (info != null && info.getTimestamp() != null) {
            refreshDisplayInfo(info);
        }
    }

    private void refreshDisplayInfo(UserInfo info) {
        DateFormat df = new SimpleDateFormat(SIMPLE_DATE_FORMAT);

        mNameText.setText(String.format("%s %s", info.getFirstName(), info.getLastName()));
        mDateOfBirthText.setText(df.format(info.getBirthDate()));

        Map<String, String> homeMunicipality = info.getHomeMunicipality();
        String languageCode = AppPreferences.getLanguageCodeSetting(mContext);
        if (homeMunicipality == null || homeMunicipality.isEmpty()) {
            mHomeMunicipality.setText(null);
        } else if (homeMunicipality.containsKey(languageCode)) {
            mHomeMunicipality.setText(homeMunicipality.get(languageCode));
        } else {
            mHomeMunicipality.setText(homeMunicipality.get(AppPreferences.LANGUAGE_CODE_FI));
        }

        if (info.getAddress() != null) {
            Address address = info.getAddress();
            mAddressText.setText(String.format(ADDRESS_FORMAT,
                    address.getStreetAddress(),
                    address.getPostalCode(),
                    address.getCity(),
                    address.getCountry() != null ? address.getCountry() : ""
            ));
        }

        refreshHunterRegistryInfo(info, df);
        refreshOccupationInfo(info, df);
    }

    private void refreshHunterRegistryInfo(UserInfo info, DateFormat df) {
        if (info.getHuntingBanStart() != null || info.getHuntingBanEnd() != null) {
            refreshHuntingBanInfo(info, df);

            mHunterIdItem.setVisibility(View.GONE);
            mPaymentItem.setVisibility(View.GONE);
            mMembershipItem.setVisibility(View.GONE);
            mInsuranceText.setVisibility(View.GONE);
            mNoValidLicenceItem.setVisibility(View.GONE);
            mHuntingBanItem.setVisibility(View.VISIBLE);
        } else if (info.getHuntingCardValidNow()) {
            refreshHuntingCardInfo(info, df);

            mHunterIdItem.setVisibility(View.VISIBLE);
            mPaymentItem.setVisibility(View.VISIBLE);
            mMembershipItem.setVisibility(View.VISIBLE);
            mInsuranceText.setVisibility(View.VISIBLE);
            mNoValidLicenceItem.setVisibility(View.GONE);
            mHuntingBanItem.setVisibility(View.GONE);
        } else {
            mHunterIdItem.setVisibility(View.GONE);
            mPaymentItem.setVisibility(View.GONE);
            mMembershipItem.setVisibility(View.GONE);
            mInsuranceText.setVisibility(View.GONE);
            mNoValidLicenceItem.setVisibility(View.VISIBLE);
            mHuntingBanItem.setVisibility(View.GONE);
        }
    }

    private void refreshHuntingBanInfo(UserInfo info, DateFormat df) {
        mHuntingBanText.setText(String.format(SIMPLE_DURATION_FORMAT,
                info.getHuntingBanStart() != null ? df.format(info.getHuntingBanStart()) : "",
                info.getHuntingBanEnd() != null ? df.format(info.getHuntingBanEnd()) : ""
        ));
    }

    private void refreshHuntingCardInfo(UserInfo info, DateFormat df) {
        mHunterIdText.setText(info.getHunterNumber());
        mPaymentText.setText(info.getHuntingCardValidNow() != null && info.getHuntingCardValidNow()
                ? String.format(getString(R.string.my_details_fee_paid_format), df.format(info.getHuntingCardStart()), df.format(info.getHuntingCardEnd()))
                : getString(R.string.my_details_fee_not_paid));

        Rhy rhy = info.getRhy();
        String languageCode = AppPreferences.getLanguageCodeSetting(mContext);
        if (rhy != null && rhy.getName() != null && !rhy.getName().isEmpty() && rhy.getName().containsKey(languageCode)) {
            mMembershipText.setText(String.format(MEMBERSHIP_NAME_FORMAT, rhy.getName().get(languageCode), rhy.getOfficialCode()));
        } else if (rhy != null && rhy.getName() != null && !rhy.getName().isEmpty()) {
            mMembershipText.setText(String.format(MEMBERSHIP_NAME_FORMAT, rhy.getName().get(AppPreferences.LANGUAGE_CODE_FI), rhy.getOfficialCode()));
        } else {
            mMembershipText.setText(null);
        }
    }

    private void refreshOccupationInfo(UserInfo info, DateFormat df) {
        mOccupationsContainer.removeAllViews();

        if (info.getOccupations() != null && !info.getOccupations().isEmpty()) {
            mOccupationsTitle.setVisibility(View.VISIBLE);
            mOccupationsContainer.removeAllViews();
            mOccupationsContainer.setVisibility(View.VISIBLE);

            LayoutInflater inflater = LayoutInflater.from(mContext);
            View separator = null;

            for (Occupation occupation : info.getOccupations()) {
                View view = inflater.inflate(R.layout.view_occupation_item, mOccupationsContainer, false);
                TextView itemText = (TextView) view.findViewById(R.id.occupation_description);
                separator = view.findViewById(R.id.occupation_separator);

                String occupationDuration;
                if (occupation.getBeginDate() == null && occupation.getEndDate() == null) {
                    occupationDuration = getText(R.string.duration_indefinite).toString();
                } else {
                    occupationDuration = String.format(SIMPLE_DURATION_FORMAT,
                            occupation.getBeginDate() != null ? df.format(occupation.getBeginDate()) : "",
                            occupation.getEndDate() != null ? df.format(occupation.getEndDate()) : ""
                    );
                }

                // Swedish rhy and occupation names are localized. Otherwise use finnish ones
                Map<String, String> occupationTitle = occupation.getName();
                Map<String, String> organizationTitle = null;
                if (occupation.getOrganisation() != null) {
                    organizationTitle = occupation.getOrganisation().getName();
                }
                String languageCode = AppPreferences.getLanguageCodeSetting(mContext);

                if (occupationTitle != null && organizationTitle != null) {
                    itemText.setText(String.format(OCCUPATION_FORMAT,
                            organizationTitle.containsKey(languageCode) ? organizationTitle.get(languageCode) : organizationTitle.get(AppPreferences.LANGUAGE_CODE_FI),
                            occupationTitle.containsKey(languageCode) ? occupationTitle.get(languageCode) : occupationTitle.get(AppPreferences.LANGUAGE_CODE_FI),
                            occupationDuration));
                } else {
                    itemText.setText(String.format(OCCUPATION_FORMAT,
                            organizationTitle != null ? organizationTitle.get(AppPreferences.LANGUAGE_CODE_FI) : null,
                            occupationTitle != null ? occupationTitle.get(AppPreferences.LANGUAGE_CODE_FI) : null,
                            occupationDuration));
                }

                mOccupationsContainer.addView(view);
            }

            // Do not display separator after last item
            if (separator != null) {
                separator.setVisibility(View.GONE);
            }
        } else {
            mOccupationsTitle.setVisibility(View.GONE);
            mOccupationsContainer.setVisibility(View.GONE);
        }
    }
}

File: riistaandroid/src/main/java/fi/riista/mobile/pages/MyDetailsFragment.kt
package fi.riista.mobile.pages

import android.content.Context
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.res.ResourcesCompat
import androidx.core.graphics.drawable.DrawableCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.Observer
import com.google.android.material.button.MaterialButton
import dagger.android.support.AndroidSupportInjection
import fi.riista.common.RiistaSDK
import fi.riista.common.huntingclub.ui.HuntingClubController
import fi.riista.common.reactive.DisposeBag
import fi.riista.common.reactive.disposeBy
import fi.riista.common.ui.controller.ViewModelLoadStatus
import fi.riista.mobile.R
import fi.riista.mobile.feature.myDetails.MyDetailsHuntingClubMembershipsFragment
import fi.riista.mobile.models.user.UserInfo
import fi.riista.mobile.repository.MetsahallitusPermitRepository
import fi.riista.mobile.riistaSdkHelpers.AppLanguageProvider
import fi.riista.mobile.riistaSdkHelpers.ContextStringProviderFactory
import fi.riista.mobile.utils.AppPreferences
import fi.riista.mobile.utils.UserInfoStore
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject

/**
 * My details/hunting card page.
 * Simply displays stored user information.
 */
class MyDetailsFragment : PageFragment() {

    @Inject
    internal lateinit var mUserInfoStore: UserInfoStore

    @Inject
    internal lateinit var mMHPermitRepository: MetsahallitusPermitRepository

    private lateinit var mContext: Context
    private var mUserInfo: UserInfo? = null
    private lateinit var mHuntingClubController: HuntingClubController
    private val disposeBag = DisposeBag()

    private lateinit var mNameText: TextView
    private lateinit var mDateOfBirthText: TextView
    private lateinit var mHomeMunicipality: TextView
    private lateinit var mAddressText: TextView

    private lateinit var mLicenseButton: Button
    private lateinit var mMetsahallitusPermitsButton: Button
    private lateinit var mShootingTestsButton: Button
    private lateinit var mOccupationsButton: Button
    private lateinit var mHuntingGroupMembershipsButton: MaterialButton

    // Dagger injection of a Fragment instance must be done in On-Attach lifecycle phase.
    override fun onAttach(context: Context) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mContext = requireContext()
        mUserInfo = mUserInfoStore.getUserInfo()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_my_details, container, false)
        setupActionBar(R.layout.actionbar_my_details, false)

        mNameText = view.findViewById(R.id.my_details_name_value)
        mDateOfBirthText = view.findViewById(R.id.my_details_date_of_birth_value)
        mHomeMunicipality = view.findViewById(R.id.my_details_home_municipality_value)
        mAddressText = view.findViewById(R.id.my_details_address_value)

        mLicenseButton = view.findViewById(R.id.my_details_license_button)
        mLicenseButton.setOnClickListener { onLicenseClick() }

        mMetsahallitusPermitsButton = view.findViewById(R.id.my_details_mh_permits_button)
        mMetsahallitusPermitsButton.visibility = View.GONE
        mMetsahallitusPermitsButton.setOnClickListener { onMetsahallitusPermitsClick() }

        mShootingTestsButton = view.findViewById(R.id.my_details_shooting_tests_button)
        mShootingTestsButton.setOnClickListener { onShootingTestsClick() }

        mOccupationsButton = view.findViewById(R.id.my_details_occupations_button)
        mOccupationsButton.setOnClickListener { onOccupationsClick() }

        mHuntingGroupMembershipsButton = view.findViewById(R.id.my_details_hunting_group_memberships_button)
        mHuntingGroupMembershipsButton.setOnClickListener { onHuntingGroupMembershipsClick() }

        mUserInfo?.username?.let { username ->
            mMHPermitRepository.findMetsahallitusPermits(username).observe(viewLifecycleOwner, Observer { list ->
                mMetsahallitusPermitsButton.visibility = if (list == null || list.isEmpty()) View.GONE else View.VISIBLE
            })
        }

        mHuntingClubController = HuntingClubController(
            huntingClubsContext = RiistaSDK.currentUserContext.huntingClubsContext,
            languageProvider = AppLanguageProvider(requireContext()),
            stringProvider = ContextStringProviderFactory.createForContext(requireContext()),
        )

        return view
    }

    override fun onResume() {
        super.onResume()

        refreshDisplayInfo()

        mHuntingClubController.viewModelLoadStatus.bindAndNotify { viewModelLoadStatus ->
            when (viewModelLoadStatus) {
                is ViewModelLoadStatus.NotLoaded,
                ViewModelLoadStatus.Loading,
                ViewModelLoadStatus.LoadFailed -> {
                }
                is ViewModelLoadStatus.Loaded -> {
                    val buttonIcon: Drawable? = if (viewModelLoadStatus.viewModel.hasOpenInvitations) {
                        AppCompatResources.getDrawable(requireContext(), R.drawable.ic_alert_circle)
                    } else {
                        null // Don't show an icon
                    }
                    buttonIcon?.let { drawable ->
                        DrawableCompat.setTint(
                            drawable,
                            ResourcesCompat.getColor(resources, R.color.traffic_light_red, null)
                        )
                    }
                    mHuntingGroupMembershipsButton.setCompoundDrawablesWithIntrinsicBounds(
                        null, null, buttonIcon, null
                    )
                }
            }
        }.disposeBy(disposeBag)

        loadHuntingClubsIfNotLoaded()
    }

    private fun loadHuntingClubsIfNotLoaded() {
        if (mHuntingClubController.viewModelLoadStatus.value is ViewModelLoadStatus.Loaded) {
            return
        }
        loadViewModel()
    }

    private fun loadViewModel(refresh: Boolean = false) {
        MainScope().launch {
            mHuntingClubController.loadViewModel(refresh)
        }
    }

    override fun onPause() {
        super.onPause()
        disposeBag.disposeAll()
    }

    private fun refreshDisplayInfo() {
        mUserInfo?.let { user ->

            mOccupationsButton.visibility = if (user.hasOccupations()) View.VISIBLE else View.GONE

            // Timestamp check is to make sure the data from server is not in outdated format.
            if (user.timestamp != null) {

                val df = SimpleDateFormat(SIMPLE_DATE_FORMAT, Locale.getDefault())
                df.timeZone = TimeZone.getTimeZone("EET")

                mNameText.text = user.firstName + " " + user.lastName
                mDateOfBirthText.text = df.format(user.birthDate)

                val homeMunicipality = user.homeMunicipality
                val languageCode = AppPreferences.getLanguageCodeSetting(mContext)

                if (homeMunicipality == null || homeMunicipality.isEmpty()) {
                    mHomeMunicipality.text = null
                } else if (homeMunicipality.containsKey(languageCode)) {
                    mHomeMunicipality.text = homeMunicipality[languageCode]
                } else {
                    mHomeMunicipality.text = homeMunicipality[AppPreferences.LANGUAGE_CODE_FI]
                }

                user.address?.let { address ->
                    mAddressText.text = String.format(ADDRESS_FORMAT,
                            address.streetAddress,
                            address.postalCode,
                            address.city,
                            address.country ?: "")
                }

                mOccupationsButton.isEnabled = user.occupations.size > 0
            }
        }
    }

    private fun onLicenseClick() {
        val fragment = MyDetailsLicenseFragment.newInstance()
        activity?.supportFragmentManager?.let { fragment.show(it, MyDetailsLicenseFragment.TAG) }
    }

    private fun onShootingTestsClick() {
        val fragment = MyDetailsShootingTestsFragment.newInstance()
        activity?.supportFragmentManager?.let { fragment.show(it, MyDetailsShootingTestsFragment.TAG) }
    }

    private fun onMetsahallitusPermitsClick() {
        mUserInfo?.let { u ->
            val fragment = MyDetailsMhPermitListFragment.newInstance(u.username)
            activity?.supportFragmentManager?.let { fragment.show(it, MyDetailsMhPermitListFragment.TAG) }
        }
    }

    private fun onOccupationsClick() {
        val fragment = MyDetailsOccupationsFragment.newInstance()
        activity?.supportFragmentManager?.let { fragment.show(it, MyDetailsOccupationsFragment.TAG) }
    }

    private fun onHuntingGroupMembershipsClick() {
        val fragment = MyDetailsHuntingClubMembershipsFragment()
        activity?.supportFragmentManager?.let { fragmentManager ->
            fragment.show(fragmentManager, MyDetailsHuntingClubMembershipsFragment.TAG)

            // Refresh viewmodel after fragment has been closed, as model might have changed
            fragmentManager.registerFragmentLifecycleCallbacks(object : FragmentManager.FragmentLifecycleCallbacks() {
                override fun onFragmentViewDestroyed(fm: FragmentManager, f: Fragment) {
                    super.onFragmentViewDestroyed(fm, f)
                    loadViewModel()
                    activity?.supportFragmentManager?.unregisterFragmentLifecycleCallbacks(this)
                }
            }, false)

        }
    }

    companion object {
        internal const val SIMPLE_DATE_FORMAT = "dd.MM.yyyy"
        internal const val SIMPLE_DURATION_FORMAT = "%s - %s"
        internal const val ADDRESS_FORMAT = "%s\n%s %s\n%s"

        @JvmStatic
        fun newInstance(): MyDetailsFragment {
            return MyDetailsFragment()
        }
    }
}
